### DISCUSSION TITLE
WIP v4

### DISCUSSION THREAD

**Discussion by visualapproach** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#428] [General]
Hi all, just wanted to let you know that I'm working on a completely new firmware, so if you want to make pull requests you may want to wait until next version is released, or it will be extra work to port it to the new version.

So, what's the plan?
Unifying all models to a single firmware. No separate builds depending on pump models.

PROs: 
* All settings are saved on the ESP so updating firmware should be a no brainer. No need to edit config files every update.
* Opens up the possibility to have an "update firmware" from the ESP itself. I don't know how to do that yet, but the possibility is there...
* In theory, we should be able to use any display with any cio. At least the 6-wire units should be interchangeable. So if you buy a spare display for example, it can be from another model. Again, this has to be tested. No promises made.

CONs:
* Since this is a massive rewrite of the code, all bells and whistles will not be there from start. Making it work at all is the main focus initially. I have made it work with very basic functions on my own pump.
* It takes a lot of time from me, and it will take some time to get it ready to release. Meantime, no updates will probably be done to current versions.
* I can't test other models than PRE2021, so you guys need to help out verifying, if you dare :-)

Take care now / Thomas

**Comment by visualapproach** [comment]
**Current status.**
Current version (offline) is working on my pump. In some ways better than before. Some issues I have to deal with is memory. I had to increase the IRAM to 48kB and shrink the icache to make it compile. I also tried ESPASyncWebserver, hoping for a speedier serving of the web pages, and it's also ESP32 compatible. Unfortunately I also had to add some other libraries to make it work and it ended up beeing a really memory heavy beast, and I haven't got it to work good enough, so back to the ESP8266Webserver.
Speaking of ESP32 - I decided I would try to port to both 8266 and 32, so I wasted a couple of days trying. Turns out LittleFS implementation is not the same so lot of work needed there if it's going to work. Plus some other things. So ESP32 is on hold for now.
There is sometimes problems reading webpages and I think it's memory related, but I have 20kB free heap so I don't know what's wrong.
I'll work on the 4-wire code now before I have something worth publishing. Lets hope we can squeeze it in, in the poor 8266.

**Reply by visualapproach** [reply]
Uploaded to development branch now. Put your feedback below

**Comment by Christians8710** [comment]
Hello, I flashed to 4.0.0 alpha on an 8266.
The websites load very slowly or not at all.
Model: S100101 6wire

![Screenshot 2023-02-05 191905](https://user-images.githubusercontent.com/118019272/216837883-5f267602-edfd-4134-a950-a7138a90a6a4.jpg)

![image](https://user-images.githubusercontent.com/118019272/216837902-a539e401-3121-46e3-a1e4-fffab59c0f96.png)

**Reply by visualapproach** [reply]
Okay. I don't know the source of this behaviour, see #439 
I wonder if it's the same on v3. Lets see what he answers on the firmware version. If it is v3 then it must be something new in one of the libraries... 
Thanks for your quick test!

**Reply by Christians8710** [reply]
If I flash the version from 2 hours ago, the following error message appears

![image](https://user-images.githubusercontent.com/118019272/218279521-e418048b-27ad-4c12-890b-d16227cead27.png)
![image](https://user-images.githubusercontent.com/118019272/218279535-3aac4322-0d87-4ed2-94d9-66d2191e8d4b.png)

**Reply by visualapproach** [reply]
> If I flash the version from 2 hours ago, the following error message appears
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/118019272/218279521-e418048b-27ad-4c12-890b-d16227cead27.png)
> 
> ![image](https://user-images.githubusercontent.com/118019272/218279535-3aac4322-0d87-4ed2-94d9-66d2191e8d4b.png)
> 
> 

It's warnings, not errors. Don't worry about them

**Reply by visualapproach** [reply]
Oh I missed the second picture with the error. Very strange, but uncomment the icache/iram line in platformio

**Comment by cobaltfish** [comment]
I'm seeing problems with 4wire

At first I got errors as above, with pages not loading.  I followed the advice on #439 and set `espressif8266@~4.1.0` and that got it running well enough to get the hardware config page up

Then when I set it to any of the 4 wire (I tried 54173, 54154, 54144 with the board disconnected from the tub) it went into restart loop.  At a network level it would be up for about 1 ping before dropping off.  I plugged it into usb and monitored and got the following (repeatedly):

```Exception (28):
epc1=0x40224053 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000

LoadProhibited: A load referenced a page mapped with an attribute that does not permit loads
  epc1=0x40224053 in DSP_4W::setStates(sStates const&) at ??:?
```
 Fuller log here:
[nodemcu_debug_2023-02-09.txt](https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/files/10695004/nodemcu_debug_2023-02-09.txt)

The only way I could get any further config was to reflash  the filesystem image to remove the settings.
I did try all the 6wire boards to try and eliminate that from the problems

For reference, the boards I'm using are these, and have been reasonably stable on the previous version: https://www.amazon.co.uk/gp/product/B07Z68HYW1/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1
I've got the v1 pcb and a TXS0108E 8 channel Logic Level Converter

**Reply by cobaltfish** [reply]
In platformio.ini I set `build_type = debug` which looks to give more helpful output:

``` 
--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Exception (28):
epc1=0x40243c98 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000

LoadProhibited: A load referenced a page mapped with an attribute that does not permit loads
  epc1=0x40243c98 in DSP_4W::setStates(sStates const&) at lib/BWC_unified/DSP_4W.cpp:93 (discriminator 2)

>>>stack>>>

ctx: cont
sp: 3fff2df0 end: 3fff3040 offset: 0190
3fff2f80:  3fffdad0 3fff091c 3fff0944 40243c79  
3fff2f90:  3fffdad0 3fff091c 3fff0910 4023dee2  
3fff2fa0:  00010100 00000000 20201463 00000020  
3fff2fb0:  00080000 000d0000 00000000 00000000  
3fff2fc0:  00000000 00000000 4025b234 402477a0  
3fff2fd0:  00000000 00000000 00000000 4022d8c3  
3fff2fe0:  00000000 00000000 00000000 00000000  
3fff2ff0:  00000000 00000000 00000000 00000000  
3fff3000:  00000000 00000000 00000000 00000000  
3fff3010:  00000000 00000000 00000000 feefeffe  
3fff3020:  feefeffe 00000000 3fffdab0 4024cf16  
3fff3030:  feefeffe feefeffe feefeffe 40102051  
<<<stack<<<

0x40243c79 in sStates::operator=(sStates const&) at lib/BWC_unified/enums.h:115
 (inlined by) DSP_4W::setStates(sStates const&) at lib/BWC_unified/DSP_4W.cpp:83
0x4023dee2 in BWC::loop() at lib/BWC_unified/bwc.cpp:181
0x4025b234 in cbuf::flush() at ??:?
0x402477a0 in EspClass::getFreeHeap() at ~/.platformio/packages/framework-arduinoespressif8266/cores/esp8266/Esp.cpp:223
0x4022d8c3 in loop at src/main.cpp:58
0x4024cf16 in loop_wrapper() at ~/.platformio/packages/framework-arduinoespressif8266/cores/esp8266/core_esp8266_main.cpp:258
0x40102051 in cont_wrapper at ~/.platformio/packages/framework-arduinoespressif8266/cores/esp8266/cont.S:81


--------------- CUT HERE FOR EXCEPTION DECODER ---------------
```
This suggests its `_to_DSP_buf[i] = rawPayloadToDSP[i];` in this codeblock in lib/BWC_unified/DSP_4W.cpp:83:

```
void DSP_4W::setStates(const sStates& to_dsp_states)
{
    _to_dsp_states = to_dsp_states;
    /* If godmode - generate payload, else send rawpayload*/
    if(to_dsp_states.godmode)
    {
        generatePayload();
    }
    else
    {
        for(int i = 0; i < PAYLOADSIZE; i++)
        {
            _to_DSP_buf[i] = rawPayloadToDSP[i];
        }
    }
}
```
and / or `    from_dsp_states = _dsp->getStates();` in lib/BWC_unified/bwc.cpp:181

Would having the board disconnected from pump and display affect it? I'll see if I can do a debug while connected to the pump and if I get other stacktrace

**Reply by visualapproach** [reply]
Great, I'll look into this! Thank you @cobaltfish !

**Reply by visualapproach** [reply]
@cobaltfish I fixed this issue, but stumbled upon another which I work on.

**Reply by visualapproach** [reply]
@cobaltfish You can try the new code now. At least it is behaving on the bench without a pump

**Reply by cobaltfish** [reply]
Its working on the bench now.  I'm using a spare board to test with, so will swap it out with the live board for more testing

**Reply by visualapproach** [reply]
Thanks, the pump cannot be off in godmode, on the bench, but I will take care of that later

**Reply by cobaltfish** [reply]
Good news: CIO is giving water temps to esp board
Bad news: E13 on display, it looks like its not getting messages through.  
I couldnt get it to go into God mode either nor schedule pump / god mode.  But I think my wifi was being a little flakey at that point, so could all be related

I did see some reboot loops and stack traces (was plugged into usb debug) but when I made a debug build I couldn't get it to happen again

**Reply by visualapproach** [reply]
Okay, you need to upload index.js and indexes.html also. Made some updates

**Reply by cobaltfish** [reply]
Apologies for the delay, I didn't get another chance over the weekend, and my dev board really wasn't wanting to work yesterday.

I've swapped the boards over and restored v3 onto one and v4 on the other and have been able to control the pump / heater / god mode via the web browser.  
However the display isn't working, I'm getting the E13.  I'm guessing its something to do with the relaying of the cio to dsp, as when I swap to my other board and reset the power, the display comes back.  I've not had a chance to look through code and compare the v3 with v4 cio code to figure out what might be going wrong

**Reply by visualapproach** [reply]
I'm glad you test it. No hurry. Latest upload is 2023-02-14-1737. I made a minor change that updates the raw payload from cio earlier. Maybe that will help. If you inspect the console in the browser, you should see TIMES/DBG = " x x x x x". Maybe that will help.
![image](https://user-images.githubusercontent.com/22987250/218804159-05931bd9-910e-420b-917d-923ba2472aa5.png)

**Reply by cobaltfish** [reply]
No joy yet, but this is the states and times debug:
{"CONTENT":"STATES","TIME":1676464421,"LCK":0,"PWR":1,"UNT":1,"AIR":0,"GRN":0,"RED":1,"FLT":1,"CH1":48,"CH2":48,"CH3":56,"HJT":0,"BRT":7,"ERR":0,"GOD":true,"TGT":10,"TMP":8,"VTF":8.09168911,"VTMC":8.106608391,"VTMF":46.59189606,"AMBC":8,"AMBF":46,"AMB":8,"VTM":8.106608391,"TGTC":10,"TMPC":8,"TGTF":50,"TMPF":46}
{"CONTENT":"TIMES","TIME":1676464421,"CLTIME":7,"FTIME":7,"UPTIME":76033,"PUMPTIME":64220,"HEATINGTIME":64793,"AIRTIME":0,"JETTIME":0,"COST":34.95210266,"FINT":30,"CLINT":14,"KWH":34.95210266,"KWHD":34.95288669,"WATT":1942,"T2R":"1.29","**DBG":" 170 2 8 0 0 10 0"**}

 {"CONTENT":"STATES","TIME":1676464488,"LCK":0,"PWR":1,"UNT":1,"AIR":0,"GRN":0,"RED":1,"FLT":1,"CH1":48,"CH2":48,"CH3":57,"HJT":0,"BRT":7,"ERR":0,"GOD":false,"TGT":10,"TMP":9,"VTF":8.09168911,"VTMC":8.134354591,"VTMF":46.64183807,"AMBC":8,"AMBF":46,"AMB":8,"VTM":8.134354591,"TGTC":10,"TMPC":9,"TGTF":50,"TMPF":48}
{"CONTENT":"TIMES","TIME":1676464488,"CLTIME":7,"FTIME":7,"UPTIME":76100,"PUMPTIME":64286,"HEATINGTIME":64859,"AIRTIME":0,"JETTIME":0,"COST":34.98770523,"FINT":30,"CLINT":14,"KWH":34.98770523,"KWHD":34.98889784,"WATT":1942,"T2R":"1.27",**"DBG":" 170 2 9 0 0 11 0"**}

I've copied your changes into the DSP_Base .cpp to debug that too with 
``` String DSP::debug()
{
    String s;
    for(int i = 0; i<7; i++) s += ' ' + String(_raw_payload_from_dsp[i]);
    return s;
} 
```
and bwc.cpp
```
    String s = _cio->debug();
    String t = _dsp->debug();
    doc["DBG"] = "cio " +s + ", dsp " + t;
```

{"CONTENT":"STATES","TIME":1676466605,"LCK":0,"PWR":1,"UNT":1,"AIR":0,"GRN":0,"RED":0,"FLT":0,"CH1":48,"CH2":49,"CH3":50,"HJT":0,"BRT":7,"ERR":0,"GOD":false,"TGT":20,"TMP":12,"VTF":24.99999809,"VTMC":24.99638748,"VTMF":76.99349976,"AMBC":20,"AMBF":68,"AMB":20,"VTM":24.99638748,"TGTC":20,"TMPC":12,"TGTF":68,"TMPF":54}
{"CONTENT":"TIMES","TIME":1676466605,"CLTIME":0,"FTIME":0,"UPTIME":126,"PUMPTIME":0,"HEATINGTIME":0,"AIRTIME":0,"JETTIME":0,"COST":0.00007,"FINT":30,"CLINT":14,"KWH":0.00007,"KWHD":0.000070476,"WATT":2,"T2R":"Already","DBG":"cio  170 2 12 0 0 14 0, dsp  0 0 0 0 0 0 0"}, 
So it looks like nothing is coming from DSP

**Reply by visualapproach** [reply]
Clever! Seems we have pinpointed the problem's whereabouts. I just uploaded a new version "2023-02-15-1718-alpha". It will increase a counter every time the ESP receives a "good" message, i.e. with the correct length and checksum.
Can you just confirm the wiring. Visit hardware config page and check these pins with your actual pins:
![image](https://user-images.githubusercontent.com/22987250/219089412-9f0d5d04-c8fc-45aa-9d81-87f5b239ff4c.png)
In this case it should be D4 and D5 but it depends on your choices of hardware. Are they possibly backwards?

**Reply by visualapproach** [reply]
Ha! actually I see now that it IS backwards! I'll fix it and upload a new version!

**Reply by visualapproach** [reply]
Updated the data/hwconfig.html file. Upload it to ESP and go to hardware config page. Adjust settings if needed and SAVE. Reboot!

**Reply by cobaltfish** [reply]
Its Working!!

Both debug and display!

{"CONTENT":"TIMES","TIME":1676553650,"CLTIME":0,"FTIME":0,"UPTIME":10946,"PUMPTIME":35,"HEATINGTIME":28,"AIRTIME":0,"JETTIME":0,"COST":0.021247778,"FINT":30,"CLINT":14,"KWH":0.021247778,"KWHD":0.021498838,"WATT":1942,"T2R":"Already","DBG":" 170 2 9 0 0 11 0msg count:149"}

Thanks so much for your help.

**Reply by visualapproach** [reply]
Cool! I think I will call it 'beta' version now that it works on both a 4 and 6-wire device.

**Comment by jamesonuk** [comment]
I am going to try V4 at some point shortly but in terms of memory and web server issues, I wonder whether it would be worthwhile looking at minifying and then gzipping all the files?   It might also be prudent to compress the images.
eg. index.html compresses down from 11kb to 3kb and wifi.html compresses down from 9kb to 2kb which I think will make a big difference
Doing a quick test and simply replacing the existing files with minified and then .gz files makes a huge different in terms of  responsiveness of web UI    I suspect compressing the images (the visualapproach png on the main page is actually much bigger than the files being served)

**Reply by jamesonuk** [reply]
Hacked together a few things I found on the net and put them together to do this at build time (zip but not minify)
https://github.com/jamesonuk/WiFi-remote-for-Bestway-Lay-Z-SPA/commit/ae8e266a5e37d1105712e1da4b844a5b940dd5b7
This seems to make a big difference in web UI performance (and only the images are now showing the content length error)
I only did this for the 6 wire code to check it works but if I get time I will attempt to do a PR for v4

**Reply by visualapproach** [reply]
I'll check it out. It's not solving the underlying cause but it's a good thing anyway! Thx @jamesonuk

**Comment by visualapproach** [comment]
I overclocked the ESP to 160 MHz. It feels quicker to serve the pages now. I updated the platformio.ini with that flag set. Maybe it helps someone. Also, thinkingn about it - I changed to a fixed IP recently. I wonder if that has something to do with the broken uploads.

**Comment by cobaltfish** [comment]
I've found a new problem.  I can't turn the filter pump off!
I did a scheduled run overnight, and it took control, turned on pump, heating, and then stopped heating when off peak electric ended, but the pump stayed on.  I tried clicking the button in the UI with the same result.  The toggle went to off, then switched back to on moments later.  The other toggles (heater, c/f, god mode) seem to work as expected, just not pump!
The debug console output shows this:
```
index.js:99 {"CONTENT":"TIMES","TIME":1676623572,"CLTIME":0,"FTIME":0,"UPTIME":80325,"PUMPTIME":36141,"HEATINGTIME":28459,"AIRTIME":0,"JETTIME":0,"COST":2.629257202,"FINT":30,"CLINT":14,"KWH":15.46621895,"KWHD":15.42931439,"WATT":42,"T2R":"17.10","DBG":" 170 2 16 0 0 18 0msg count:134361"}
index.js:101 {CONTENT: 'TIMES', TIME: 1676623572, CLTIME: 0, FTIME: 0, UPTIME: 80325, …}
index.js:99 {"CONTENT":"OTHER","MQTT":-1,"HASJETS":true,"HASGOD":true,"MODEL":"NO54144","RSSI":-41,"IP":"192.168.0.30","SSID":"<Wifi-SSID>","FW":"2023-02-15-1718-alpha"}
index.js:101 {CONTENT: 'OTHER', MQTT: -1, HASJETS: true, HASGOD: true, MODEL: 'NO54144', …}
index.js:253 togglePump
index.js:254 number
index.js:307 {"CMD":4,"VALUE":false,"XTIME":1676623573,"INTERVAL":0,"TXT":""}
index.js:99 {"CONTENT":"STATES","TIME":1676623574,"LCK":0,"PWR":1,"UNT":1,"AIR":0,"GRN":0,"RED":0,"FLT":1,"CH1":48,"CH2":49,"CH3":54,"HJT":0,"BRT":8,"ERR":0,"GOD":true,"TGT":37,"TMP":16,"VTF":15,"VTMC":15.01000023,"VTMF":59.01800156,"AMBC":11,"AMBF":52,"AMB":11,"VTM":15.01000023,"TGTC":37,"TMPC":16,"TGTF":99,"TMPF":61}
index.js:101 {CONTENT: 'STATES', TIME: 1676623574, LCK: 0, PWR: 1, UNT: 1, …}
index.js:99 {"CONTENT":"TIMES","TIME":1676623574,"CLTIME":0,"FTIME":0,"UPTIME":80327,"PUMPTIME":36143,"HEATINGTIME":28459,"AIRTIME":0,"JETTIME":0,"COST":2.629261255,"FINT":30,"CLINT":14,"KWH":15.46624184,"KWHD":15.42933773,"WATT":42,"T2R":"17.10","DBG":" 170 2 16 0 0 18 0msg count:134365"}
index.js:101 {CONTENT: 'TIMES', TIME: 1676623574, CLTIME: 0, FTIME: 0, UPTIME: 80327, …}
```
I've not spotted anything obvious in the web pages or c++ source code.
As best I can tell, the same code is used to generate the toggle type commands in the web pages, and the heater and hydrojet toggles work as expected.  (Toggling the hydrojets is a work around, as that stops the pump as expected)
The jump tables and masks are the same between v3 and v4, so I  don't think its there
I did wonder if it was a boolean vs integer mapping problem problem as the command value is true/false, but I states for pump (FLT) is 0/1
But hydrojets send true/false via web page, and work correctly, and the command entry to stop the pump has 0 not false

Is there any more debugging you would like me to do?

**Reply by visualapproach** [reply]
Hi! My guess it's got something to do with the delayed pump off feature. I had to do it differently in this version. The point of the feature is to continue pumping until the heater elements has cooled down a little.

**Reply by visualapproach** [reply]
Yep, I added a couple of missing lines. An if statement without an else that catched the descibed situation :-)

**Reply by cobaltfish** [reply]
Yep. thats working nicely now.  
Something is seeming a little fussy, as it kept rebooting (with stack traces when I put serial monitor) when I first built / deployed and then took control.  I then made a debug build to trace it properly, and of course, its working perfectly!

**Reply by visualapproach** [reply]
hmm, let me know if you find out anything.

**Reply by cobaltfish** [reply]
I've found something different.  On bench with 4 or 6 wire, and on the unit itself, enabling mqtt is causing it to crash and reboot
I don't think it is my config differing, as I've got the same target server, and the log entries look similar.  
The stacktrace is quite big, so I've attached it, but I think relevant bit is this
```
...
0x402337f6 in PubSubClient::write(unsigned char, unsigned char*, unsigned short) at .pio/libdeps/nodemcuv2/PubSubClient/src/PubSubClient.cpp:599
0x40233994 in PubSubClient::publish(char const*, unsigned char const*, unsigned int, bool) at .pio/libdeps/nodemcuv2/PubSubClient/src/PubSubClient.cpp:469
0x40257658 in ArduinoJson6201_F1::SafeBoolIdom<ArduinoJson6201_F1::JsonString>::safeBoolHelper() const at .pio/libdeps/nodemcuv2/ArduinoJson/src/ArduinoJson/Misc/SafeBoolIdiom.hpp:15
0x402339e1 in PubSubClient::publish(char const*, char const*, bool) at .pio/libdeps/nodemcuv2/PubSubClient/src/PubSubClient.cpp:442 (discriminator 4)
0x402a1a7c in etharp_output at ??:?
0x4021aca4 in setupHA() at src/ha.txt:47
0x40102970 in pp_soft_wdt_feed_local at ??:?
...
```

based on `0x4021aca4 in setupHA() at src/ha.txt:47` it looks like it points at line 47 of ha.txt `mqttClient.loop();`
I've not tried investigating / fixing the code in there, but instead commented out the `setupHA()` in main.cpp, and it keeps running, with mqtt data coming through

[mqtt-serial-debug.txt](https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/files/10766551/mqtt-serial-debug.txt)

**Reply by visualapproach** [reply]
Are you building in debug mode? I get an error 1 when I use debug mode...

**Reply by visualapproach** [reply]
I can't replicate the problem. I added a check for mqtt connection in setupHA(), in case that was the problem.
Mqtt runs fine without this function, it's just setting up the home assistant auto discovery.

**Reply by cobaltfish** [reply]
ugh, I have no idea.  I have reset back to branch, recompiled (non debug) and its working! I think we can safely ignore my comment earlier.  Sorry for my false alarm

**Comment by deeja** [comment]
My setup is the one here https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/discussions/383
Downloaded `v4.1.3` and also tried the OTA `master` and `beta` firmware.
- PCB: custom
- 2021 for the display and pump. 

![image](https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/assets/174247/a0d004e7-ddae-4dce-9b26-2e8e6b59d3ea)

I get no output on the screen; no start noise. 

I've reverted to v3.4 and have no issues with the following modifications:
```cpp
//Code/6-wire-version/lib/BWC/model.h b/Code/6-wire-version/lib/BWC/model.h
#define MIAMI2021  //no hydrojets
```
```cpp
//Code/6-wire-version/src/main.cpp
bwc.begin(D3,D2,D1,D5,D6,D7,D8 );
``` 

On a side note, the 2 metre extension with the shielded cable works now that I have grounded one end.

**Reply by deeja** [reply]
I put some logging in the code, and it looks like the pins are set correctly. 
Maybe it's something not pin related?
``` 
cio_data_td_pin = D3
cio_clk_pin = D2
cio_cs_ld_pin = D1
dsp_data_td_pin = D5
dsp_clk_pin = D6
dsp_cs_ld_pin = D7
dsp_audio_pi = D8
```

**Reply by visualapproach** [reply]
I wonder if it's a power thing with v4. It is overclocked to 160 MHz and maybe that's enough for the outputs to be slightly lower output. No idea really but thought I'd share my thoughts.

**Reply by deeja** [reply]
I can check next time I've pulled it apart

**Reply by visualapproach** [reply]
In dev branch 2023-10-22 there is a new page layzspa.local/inputs/ where you can check on what pins the TX lines are connected.

**Comment by visualapproach** [comment]
The UI UPDATE function stopped working. Possibly by some change on the github server. I have now spent several days on my vacation to fix that and other bugs introduced when freeing up memory for the updater to work. But now I feel it should be much more stable. It will not leave 0-size files behind, because it first downloads to a temp file "dl" and if that is bigger than zero bytes it will copy it to the proper file. Not watertight but much better than before. In the process of getting this to work there is much more memory available for the rest of the program which will be less prone to mysterious memory errors.

v4.2.0 is uploaded to both beta and master branches. You'll need to update (once) via ordinary OTA or USB if your UI update page is not working.
